Design.md

```diff
@@ -0,0 +1,175 @@
+Job worker service that provides an API to run arbitrary Linux processes.
+
+## Details
+
+
+### Security
+Use mTLS for gRPC communication.
+
+Required files:
+| File | Description |
+| --- | --- |
+| server_cert.pem | the server's certificate (public key) |
+| server_key.pem | the server's private key |
+| server_ca_cert.pem | the certificate of the CA that can verify the server's certificate. |
+| client_cert.pem | the client's certificate (public key). |
+| client_key.pem | the client's private key. |
+| client_ca_cert.pem | the certificate of the CA that can verify the client's certificate |
+
+Use `openssl` to generate all of the required certs and keys mentioned above.
+
+User authentication and authorization
+
+Simple authorization:<br />
+Uses hardcoded user 'tokens' to identify known users.
+use gRPC interceptors to check if the user is authorized.
+* On client: add user token
+* On server: check the user token against known, reject connection if unknown
+TODO add actual user authentication
+
+TODO need a way to select the current user: 
+Options:
+command line argument: `... -u alex`
+cli command: `jws login <user_name>`
+  user name would be stored in an env file and read in at startup
+
+### Privacy
+Users can only interact with jobs that they started.
+
+### UX - CLI
+Scenarios<br />
+Start a job:<br />
+caller provides the command and arguments in a single string *** explore options 
+`jws start -c "ls -l" -u alex`<br />
+returns the unique identifier of the job 
+
+Stop a job:<br />
+`jws stop -j <job_id> -u alex`<br />
+returns an error if something went wrong (unknown user, invalid job_id, access denied)
+
+Get the status of a job:<br />
+`jws status -j <job_id>`<br />
+returns the status of the job: "Pending, Active, Terminated"
+  or an error: (uknown user, invalid job_id, access denied)
+
+Get the output of a job:<br />
+`jws output -j <job_id>`<br />
+
+<br />
+
+Sample usage:<br />
+`jws start -c "ping google" -u alex`<br />
+returns 100
+
+`jws status -j 100 -u alex`<br />
+returns Active
+
+`jws status -j 100 -u betty`<br />
+returns Access Denied
+
+`jws output -j 100 -u alex`<br />
+returns:
+...
+
+from a second terminal:<br />
+`jsw stop -j 100 -u alex`<br />
+returns success (no error)
+
+in the previous terminal:<br />
+`jws output -j 100 -u alex`<br />
+sees:
+... <insert output here>
+
+
+### Proto Specifications
+```
+service JobWorkerService {
+  rpc StartJob(StartJobRequest) returns (StartJobResponse);
+  rpc StopJob(StopJobRequest) returns (StopJobResponse);
+  rpc GetJobStatus(GetJobStatusRequest) returns (GetJobStatusResponse);
+  rpc GetJobOutput(GetJobOutputRequest) returns (stream GetJobOutputResponse);
+}
+
+message StartJobRequest {
+  string cmd = 1;
+}
+
+message StartJobResponse {
+  string job_id = 1;
+}
+
+message StopJobRequest {
+  string job_id = 1;
+}
+
+message StopJobResponse {}
+
+message GetJobStatusRequest {
+  string job_id = 1;
+}
+
+message GetJobStatusResponse {
+  string status = 1;
```

Comment:
Could a proto enum be used rather than a string ? This gives much more type safety.

Reply:
Will use an enum.